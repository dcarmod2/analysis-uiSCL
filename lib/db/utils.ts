import mapValues from 'lodash/mapValues'
import {ObjectID} from 'mongodb'

export function serializeCollection(obj) {
  if (obj instanceof Date) return obj.toISOString()
  if (obj instanceof ObjectID) return obj.toString()
  if (Array.isArray(obj)) return obj.map(serializeCollection)
  if (typeof obj === 'object') return mapValues(obj, serializeCollection)
  return obj
}
