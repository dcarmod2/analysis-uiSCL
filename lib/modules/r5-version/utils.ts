import {MINIMUM_R5_VERSION, VERSION_PARSE_REGEX} from './constants'

/**
 * Convert an r5 version to a number for comparing
 */
export function versionToNumber(version: string): number {
  const versionMatch = (version || '').match(VERSION_PARSE_REGEX)
  if (versionMatch) {
    const [, major, minor, patch, commit] = versionMatch

    // NB Assumes no major, minor, patch, or commit version exceeds 100
    return (
      parseInt(major) * 10000 +
      parseInt(minor) * 100 +
      parseInt(patch) +
      parseInt(commit || '0') / 100
    )
  }

  return 0
}

/**
 * Check if the r5 version is older than a given minimum.
 */
export function workerVersionInRange(
  version: string,
  min: string = MINIMUM_R5_VERSION,
  max = 'v1000.0.0'
): boolean {
  const versionNumber = versionToNumber(version)
  return (
    versionNumber >= versionToNumber(min) &&
    versionNumber <= versionToNumber(max)
  )
}

/**
 * Check if the r5 version is a patch version or in a given range
 */
export function workerVersionTestOrInRange(
  version: string,
  min: string,
  max?: string
): boolean {
  const isTestVersion = versionToNumber(version) % 1 !== 0
  return isTestVersion || workerVersionInRange(version, min, max)
}
