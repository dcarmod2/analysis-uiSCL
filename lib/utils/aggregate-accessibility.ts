import get from 'lodash/get'
import range from 'lodash/range'

const N_HISTOGRAM_BINS = 15

/**
 * Compute the accessibility, exported for use in comparison weighted average
 * accessibility.
 * @param weights a grid containing the weights (e.g. population)
 * @param aggregationArea a grid containing the aggregation area, with the
 *        percentage the aggregation area overlaps each cell, scaled between 0
 *        (no overlap) and 100,000 (full coverage)
 * @param accessibility a grid containing the accessibility (i.e. the regional
 *        analysis results)
 */
export default function getAggregateAccessibility(
  accessibility,
  aggregationArea,
  weights
) {
  const aa = get(aggregationArea, 'grid')
  if (weights == null || aa == null || accessibility == null) {
    return
  }

  if (accessibility.zoom !== weights.zoom || accessibility.zoom !== aa.zoom) {
    console.error('Zoom levels of all grids in weighted average must match!')
  }

  // find max and min value for histogram
  let maxAccessibility = 0
  let minAccessibility = Infinity

  // Pass one: find min and max from aggregationArea
  // aggregationArea is probably the smallest grid and is guaranteed to contain
  // all pixels
  for (let y = 0, pixel = 0; y < aa.height; y++) {
    // Transform to coordinates in other grid; all grids have same zoom, so
    // translate based on the difference in where the edges of the grid are
    const accessY = y + aa.north - accessibility.north
    for (let x = 0; x < aa.width; x++, pixel++) {
      if (aa.data[pixel] > 0) {
        const accessX = x + aa.west - accessibility.west
        const accessibilityThisCell = accessibility.contains(accessX, accessY)
          ? accessibility.data[accessY * accessibility.width + accessX]
          : 0

        minAccessibility = Math.min(accessibilityThisCell, minAccessibility)
        maxAccessibility = Math.max(accessibilityThisCell, maxAccessibility)
      }
    }
  }

  const spread = maxAccessibility - minAccessibility
  const bins = range(N_HISTOGRAM_BINS).map(() => 0)

  let sumOfWeights = 0
  // weight and accessibility value for each cell, used to compute percentiles
  const weightedAccessibility = []

  // pass two: accumulate into bins
  for (let y = 0, pixel = 0; y < aa.height; y++) {
    const accessY = y + aa.north - accessibility.north
    const weightY = y + aa.north - weights.north
    for (let x = 0; x < aa.width; x++, pixel++) {
      if (aa.data[pixel] > 0) {
        const accessX = x + aa.west - accessibility.west
        const accessibilityThisCell = accessibility.contains(accessX, accessY)
          ? accessibility.data[accessY * accessibility.width + accessX]
          : 0

        let binIndex = Math.floor(
          ((accessibilityThisCell - minAccessibility) / spread) *
            N_HISTOGRAM_BINS
        )

        if (binIndex > N_HISTOGRAM_BINS) console.error('invalid bin!')

        // avoid putting the highest accessibility cell beyond the histogram
        if (binIndex === N_HISTOGRAM_BINS) binIndex--

        // how much of this cell is within the aggregationArea
        const aggregationAreaWeightThisCell = aa.data[pixel] / 100000

        const weightX = x + aa.west - weights.west
        const weightThisCell = weights.contains(weightX, weightY)
          ? weights.data[weightY * weights.width + weightX]
          : 0

        const weight = aggregationAreaWeightThisCell * weightThisCell
        bins[binIndex] += aggregationAreaWeightThisCell * weightThisCell

        sumOfWeights += weight
        weightedAccessibility.push({
          weight,
          accessibility: accessibilityThisCell
        })
      }
    }
  }

  // Figure out percentiles
  let cumulativeWeight = 0
  // Preinitialize with null so 50th percentile is percentiles[50].
  // There is no zeroth percentile
  const percentiles = [0]
  const weightPerPercentile = sumOfWeights / 100
  let expectedWeight = weightPerPercentile
  weightedAccessibility.sort((a, b) => a.accessibility - b.accessibility)

  // we calculate and return an old-style weighted average for comparison purposes.
  let sumOfWeightedAccessibility = 0

  // We loop over the sorted weighted accessibility values. When we cross a
  // percentile boundary, we push that percentile onto the list
  weightedAccessibility.forEach(({accessibility, weight}) => {
    sumOfWeightedAccessibility += accessibility * weight

    cumulativeWeight += weight

    // while not if, possible to cross many percentiles with one high-density cell
    while (cumulativeWeight > expectedWeight) {
      // There is a percentile value somewhere within the values at this sample
      // We use > above, not >=, so that a percentile value that falls exactly
      // between two accessibility values will get the higher value, e.g. if 10
      // people have accessibility 50 and ten people have accessibility 60, we
      // want 60. This is because in the UI we say "x% of <people> have
      // accessibility y or higher," so we want y to be as high as possible
      // while still being a true statement.
      percentiles.push(accessibility)
      expectedWeight += weightPerPercentile
    }
  })

  // includes values 1 - 99
  // When the resolution of the data is low the last item may only define (say)
  // the 90th percentile, all remaining percentiles are just the max. Also
  // handles the case when there is no accessibility within the aggregation
  // area by returning all zeros.
  while (percentiles.length < 99) {
    percentiles.push(
      weightedAccessibility.length > 0
        ? weightedAccessibility[weightedAccessibility.length - 1].accessibility
        : 0
    )
  }

  const binWidth = spread / N_HISTOGRAM_BINS

  return {
    minAccessibility,
    maxAccessibility,
    percentiles,
    weightedAverage: sumOfWeightedAccessibility / sumOfWeights,
    bins: bins.map((value, idx) => ({
      value,
      min: minAccessibility + idx * binWidth,
      max: minAccessibility + (idx + 1) * binWidth
    }))
  }
}
