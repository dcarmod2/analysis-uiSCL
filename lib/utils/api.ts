// Query values are typed as string OR string[]
export function getQueryAsString(p: string | string[]): string {
  return Array.isArray(p) ? p[0] : p
}

// Get a query as an object
export function getQueryAsObject(
  p: string | string[]
): Record<string, unknown> {
  try {
    return JSON.parse(getQueryAsString(p))
  } catch (e) {
    return {}
  }
}

// Convert an error to POJO for sending over the wire
export function errorToPOJO(e: Error): Record<string, unknown> {
  const obj = {}
  Object.getOwnPropertyNames(e).forEach((k) => {
    obj[k] = e[k]
  })
  return obj
}
