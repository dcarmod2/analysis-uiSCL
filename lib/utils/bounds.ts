import lonlat from '@conveyal/lonlat'

import L from 'lib/leaflet'
import message from 'lib/message'

import reprojectCoordinates from './reproject-coordinates'

const ANTIMERIDIAN_WARNING =
  'We do not currently support bounds that cross the antimeridian.'

type Bounds = {
  north: number
  south: number
  east: number
  west: number
  toString: () => string
}

export function toLatLngBounds(
  bounds: Bounds | L.LatLngBounds
): L.LatLngBounds {
  // Check if it's already a leaflet bounds object
  if (bounds instanceof L.LatLngBounds) return bounds
  return new L.LatLngBounds([
    [bounds.north, bounds.west],
    [bounds.south, bounds.east]
  ])
}

/**
 * NB: We do not currently support crossing the antimeridian.
 */
export function fromLatLngBounds(bounds: L.LatLngBounds) {
  // Don't change the original bounds object
  let b = new L.LatLngBounds(bounds.getSouthWest(), bounds.getNorthEast())

  if (b.getWest() < -180) {
    b = new L.LatLngBounds(new L.LatLng(b.getSouth(), -180), b.getNorthEast())
    window.alert(message('antimeridian', ANTIMERIDIAN_WARNING))
  }

  if (b.getEast() > 180) {
    b = new L.LatLngBounds(new L.LatLng(b.getSouth(), 180), b.getNorthEast())
    window.alert(message('antimeridian', ANTIMERIDIAN_WARNING))
  }

  const bboxString = b.toBBoxString()
  return {
    north: b.getNorth(),
    south: b.getSouth(),
    east: b.getEast(),
    west: b.getWest(),
    toString: () => bboxString
  }
}

/**
 * Reproject bounds to the grid size that we use.
 */
export function reprojectBounds(bounds) {
  return new L.LatLngBounds(
    lonlat.toLeaflet(reprojectCoordinates(bounds.getSouthWest())),
    lonlat.toLeaflet(reprojectCoordinates(bounds.getNorthEast()))
  )
}
