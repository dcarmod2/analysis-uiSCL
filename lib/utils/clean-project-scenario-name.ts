//
import message from 'lib/message'

import {UNDEFINED_PROJECT_NAME} from '../constants'

export default function cleanProjectScenarioName(project, scenarioIndex) {
  if (!project) {
    return UNDEFINED_PROJECT_NAME
  }
  const scenarioName =
    project.variants[scenarioIndex] || message('variant.baseline')
  const description = `${project.name}-${scenarioName}`
  return description.replace(/[^a-zA-Z0-9]/, '-')
}
