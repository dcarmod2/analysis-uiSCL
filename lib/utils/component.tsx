import {ThemeProvider} from '@chakra-ui/core'
import {expect} from '@jest/globals'
import enzyme from 'enzyme'
import {Router} from 'next/router'
import {RouterContext} from 'next/dist/next-server/lib/router-context'
import React from 'react'
import {Map} from 'react-leaflet'
import {Provider} from 'react-redux'

import {makeMockStore, mockStores} from './mock-data'

// Next.js uses Math.random to generate unique ids for pages that get captured in snapshots.
global.Math.random = () => 0.5

// Set __NEXT_DATA__ (required by Router)
window.__NEXT_DATA__ = {} as any
const router = new Router('page', {}, 'as-path', {} as any)

const getName = (c) => c.displayName || c.name

export function testComponent(
  TestComponent,
  props = {},
  initialData = mockStores.init
) {
  const store = makeMockStore(initialData)
  const Wrapping = (props) => (
    <Provider store={store}>
      <ThemeProvider>
        <RouterContext.Provider value={router}>
          {props.children}
        </RouterContext.Provider>
      </ThemeProvider>
    </Provider>
  )

  const mount = () =>
    enzyme.mount(<TestComponent {...props} />, {
      wrappingComponent: Wrapping
    })

  // TODO: Dive down through the `TestComponent` automatically
  const shallow = () =>
    enzyme.shallow(<TestComponent {...props} />, {
      wrappingComponent: Wrapping
    })

  // For debugging purposes (this is not the original component!)
  const name = `WithContext(${getName(TestComponent)})`

  return {mount, name, router, shallow, store}
}

/**
 * Rename to `wrapComponent`. TODO replace all `testComponent`
 */
export const wrapComponent = testComponent

/**
 * Helper function to create quick snapshot tests for components.
 */
export function testAndSnapshot(C, p, id?: any) {
  test(`Component ${getName(C)} snapshot(mount)`, () => {
    const c = testComponent(C, p, id)
    const t = c.mount()
    expect(t).toMatchSnapshot()
    t.unmount()
  })
}

/**
 * Test and snapshot map component
 */
export function wrapMapComponent(C, p, id?: any) {
  const MapWrapper = (props) => (
    <Map>
      <C {...props} />
    </Map>
  )
  MapWrapper.displayName = `WithMap(${getName(C)})`
  return testComponent(MapWrapper, p, id)
}

function snapshotTest(wc, type) {
  test(`Component ${getName(wc)} snapshot(${type})`, () => {
    const s = wc[`${type}`]()
    expect(s).toMatchSnapshot()
    s.unmount()
  })
}

/**
 * Shallow snapshot test.
 */
export function shallowSnapshot(wc) {
  snapshotTest(wc, 'shallow')
}

/**
 * Deep snapshot test.
 */
export function deepSnapshot(wc) {
  snapshotTest(wc, 'mount')
}
