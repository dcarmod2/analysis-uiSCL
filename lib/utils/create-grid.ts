/**
 * Create a Grid from an ArrayBuffer retrieved from the server or S3. Binary
 * data that uses a common header format used across Conveyal libraries.
 */
export default function create(data: ArrayBuffer): CL.ParsedGrid {
  const array = new Int32Array(data, 4 * 5)
  const header = new Int32Array(data)

  let min = Infinity
  let max = -Infinity

  for (let i = 0, prev = 0; i < array.length; i++) {
    array[i] = prev += array[i] // de-delta code
    if (prev < min) min = prev
    if (prev > max) max = prev
  }

  const width = header[3]
  const height = header[4]

  function contains(x: number, y: number) {
    return x >= 0 && x < width && y >= 0 && y < height
  }

  // Return 0 for cells outside of the grid
  function getValue(x: number, y: number) {
    if (contains(x, y)) return array[y * width + x]
    return 0
  }

  // parse header
  return {
    zoom: header[0],
    west: header[1],
    north: header[2],
    width,
    height,
    data: array,
    min,
    max,
    contains,
    getValue
  }
}
