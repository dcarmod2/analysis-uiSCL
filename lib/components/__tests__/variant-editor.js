import {testComponent} from 'lib/utils/component'

import VariantEditor from '../variant-editor'

describe('Component > VariantEditor', () => {
  it('renders correctly', () => {
    const props = {
      variants: [],
      createVariant: jest.fn(),
      deleteVariant: jest.fn(),
      editVariantName: jest.fn(),
      showVariant: jest.fn()
    }

    // mount component
    const p = testComponent(VariantEditor, props)
    const tree = p.mount()
    expect(tree).toMatchSnapshot()
    const noCalls = [
      'createVariant',
      'deleteVariant',
      'editVariantName',
      'showVariant'
    ]
    noCalls.forEach((fn) => {
      expect(props[fn]).not.toHaveBeenCalled()
    })
    tree.unmount()
  })
})
