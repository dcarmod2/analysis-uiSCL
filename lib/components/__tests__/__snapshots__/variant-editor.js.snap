// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component > VariantEditor renders correctly 1`] = `
<Memo(Variants)
  createVariant={[MockFunction]}
  deleteVariant={[MockFunction]}
  editVariantName={[MockFunction]}
  showVariant={[MockFunction]}
  variants={Array []}
>
  <Box
    p={2}
  >
    <div
      className="css-olyig7"
    >
      <Button
        isFullWidth={true}
        leftIcon="small-add"
        onClick={[Function]}
        variantColor="green"
      >
        <PseudoBox
          _active={
            Object {
              "bg": "green.700",
            }
          }
          _disabled={
            Object {
              "boxShadow": "none",
              "cursor": "not-allowed",
              "opacity": "40%",
            }
          }
          _focus={
            Object {
              "boxShadow": "outline",
            }
          }
          _hover={
            Object {
              "bg": "green.600",
            }
          }
          alignItems="center"
          appearance="none"
          as="button"
          bg="green.500"
          borderRadius="md"
          color="white"
          display="inline-flex"
          fontSize="md"
          fontWeight="semibold"
          height={10}
          justifyContent="center"
          lineHeight="1.2"
          minWidth={10}
          onClick={[Function]}
          outline="none"
          position="relative"
          px={4}
          transition="all 250ms"
          type="button"
          userSelect="none"
          verticalAlign="middle"
          whiteSpace="nowrap"
          width="full"
        >
          <button
            className="css-sci538"
            onClick={[Function]}
            type="button"
          >
            <ButtonIcon
              icon="small-add"
              ml={-1}
              mr={2}
            >
              <Icon
                aria-hidden="true"
                color="currentColor"
                focusable="false"
                ml={-1}
                mr={2}
                name="small-add"
              >
                <Styled(div)
                  aria-hidden="true"
                  as="svg"
                  color="currentColor"
                  display="inline-block"
                  focusable="false"
                  ml={-1}
                  mr={2}
                  role="presentation"
                  size="1em"
                  verticalAlign="middle"
                  viewBox="0 0 20 20"
                >
                  <svg
                    aria-hidden="true"
                    className="css-yxiis9"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 20 20"
                  >
                    <path
                      d="M14 9h-3V6c0-.55-.45-1-1-1s-1 .45-1 1v3H6c-.55 0-1 .45-1 1s.45 1 1 1h3v3c0 .55.45 1 1 1s1-.45 1-1v-3h3c.55 0 1-.45 1-1s-.45-1-1-1z"
                      fill="currentColor"
                      fillRule="evenodd"
                    />
                  </svg>
                </Styled(div)>
              </Icon>
            </ButtonIcon>
            Create a scenario
          </button>
        </PseudoBox>
      </Button>
    </div>
  </Box>
  <ForwardRef
    py={2}
  >
    <Flex
      direction="column"
      py={2}
    >
      <Box
        display="flex"
        flexDirection="column"
        py={2}
      >
        <div
          className="css-17oo1d2"
        >
          <Text
            key=".0"
            mb={2}
            px={4}
            py={2}
          >
            <Box
              as="p"
              fontFamily="body"
              mb={2}
              px={4}
              py={2}
            >
              <p
                className="css-srcs0l"
              >
                A Scenario is a collection of modifications within a project.
              </p>
            </Box>
          </Text>
          <ForwardRef
            key=".1"
            mb={2}
            mx={4}
          >
            <Box
              as="hr"
              border="0"
              borderBottom="0.0625rem solid"
              borderColor="inherit"
              mb={2}
              mx={4}
              my={2}
              opacity="0.6"
              width="auto"
            >
              <hr
                className="css-1s2m7jw"
              />
            </Box>
          </ForwardRef>
          <InnerDock
            key=".2"
            mb={null}
          >
            <SimpleBar
              className="InnerDock "
              style={
                Object {
                  "height": 768,
                }
              }
            >
              <div
                className="InnerDock "
                data-simplebar={true}
                style={
                  Object {
                    "height": 768,
                  }
                }
              >
                <div
                  className="simplebar-wrapper"
                >
                  <div
                    className="simplebar-height-auto-observer-wrapper"
                  >
                    <div
                      className="simplebar-height-auto-observer"
                    />
                  </div>
                  <div
                    className="simplebar-mask"
                  >
                    <div
                      className="simplebar-offset"
                    >
                      <div
                        className="simplebar-content-wrapper"
                      >
                        <div
                          className="simplebar-content"
                        >
                          <ForwardRef
                            id="scenarios"
                            pl={4}
                            pr={2}
                            pt={2}
                            spacing={3}
                          >
                            <Flex
                              direction="column"
                              id="scenarios"
                              pl={4}
                              pr={2}
                              pt={2}
                            >
                              <Box
                                display="flex"
                                flexDirection="column"
                                id="scenarios"
                                pl={4}
                                pr={2}
                                pt={2}
                              >
                                <div
                                  className="css-1pygc6s"
                                  id="scenarios"
                                >
                                  <Flex
                                    key=".0"
                                    mb={null}
                                    pr={2}
                                  >
                                    <Box
                                      display="flex"
                                      mb={null}
                                      pr={2}
                                    >
                                      <div
                                        className="css-1pjcatt"
                                      >
                                        <Text
                                          flex="1"
                                          fontWeight="bold"
                                        >
                                          <Box
                                            as="p"
                                            flex="1"
                                            fontFamily="body"
                                            fontWeight="bold"
                                          >
                                            <p
                                              className="css-13c2z23"
                                            >
                                              Baseline
                                            </p>
                                          </Box>
                                        </Text>
                                        <Tip
                                          label="Baseline (empty scenario) cannot be modified"
                                        >
                                          <Tooltip
                                            aria-label="Tooltip: Baseline (empty scenario) cannot be modified"
                                            fontSize="md"
                                            hasArrow={true}
                                            label="Baseline (empty scenario) cannot be modified"
                                            zIndex={1500}
                                          >
                                            <Icon
                                              mr={1}
                                              name="lock"
                                              onBlur={[Function]}
                                              onClick={[Function]}
                                              onFocus={[Function]}
                                              onMouseEnter={[Function]}
                                              onMouseLeave={[Function]}
                                            >
                                              <Styled(div)
                                                as="svg"
                                                color="currentColor"
                                                display="inline-block"
                                                focusable={false}
                                                mr={1}
                                                onBlur={[Function]}
                                                onClick={[Function]}
                                                onFocus={[Function]}
                                                onMouseEnter={[Function]}
                                                onMouseLeave={[Function]}
                                                role="presentation"
                                                size="1em"
                                                verticalAlign="middle"
                                                viewBox="0 0 24 24"
                                              >
                                                <svg
                                                  className="css-11vxqx9"
                                                  focusable={false}
                                                  onBlur={[Function]}
                                                  onClick={[Function]}
                                                  onFocus={[Function]}
                                                  onMouseEnter={[Function]}
                                                  onMouseLeave={[Function]}
                                                  role="presentation"
                                                  viewBox="0 0 24 24"
                                                >
                                                  <path
                                                    d="M19.5,9.5h-.75V6.75a6.75,6.75,0,0,0-13.5,0V9.5H4.5a2,2,0,0,0-2,2V22a2,2,0,0,0,2,2h15a2,2,0,0,0,2-2V11.5A2,2,0,0,0,19.5,9.5Zm-9.5,6a2,2,0,1,1,3,1.723V19.5a1,1,0,0,1-2,0V17.223A1.994,1.994,0,0,1,10,15.5ZM7.75,6.75a4.25,4.25,0,0,1,8.5,0V9a.5.5,0,0,1-.5.5H8.25a.5.5,0,0,1-.5-.5Z"
                                                    fill="currentColor"
                                                  />
                                                </svg>
                                              </Styled(div)>
                                            </Icon>
                                            <ForwardRef
                                              anchorEl={
                                                <svg
                                                  class="css-11vxqx9"
                                                  focusable="false"
                                                  role="presentation"
                                                  viewBox="0 0 24 24"
                                                >
                                                  <path
                                                    d="M19.5,9.5h-.75V6.75a6.75,6.75,0,0,0-13.5,0V9.5H4.5a2,2,0,0,0-2,2V22a2,2,0,0,0,2,2h15a2,2,0,0,0,2-2V11.5A2,2,0,0,0,19.5,9.5Zm-9.5,6a2,2,0,1,1,3,1.723V19.5a1,1,0,0,1-2,0V17.223A1.994,1.994,0,0,1,10,15.5ZM7.75,6.75a4.25,4.25,0,0,1,8.5,0V9a.5.5,0,0,1-.5.5H8.25a.5.5,0,0,1-.5-.5Z"
                                                    fill="currentColor"
                                                  />
                                                </svg>
                                              }
                                              arrowSize="10px"
                                              bg="gray.700"
                                              borderRadius="sm"
                                              color="whiteAlpha.900"
                                              fontSize="md"
                                              fontWeight="medium"
                                              hasArrow={true}
                                              isOpen={false}
                                              maxW="320px"
                                              modifiers={
                                                Object {
                                                  "offset": Object {
                                                    "enabled": true,
                                                    "offset": "0, 8",
                                                  },
                                                }
                                              }
                                              placement="auto"
                                              pointerEvents="none"
                                              px="8px"
                                              py="2px"
                                              shadow="md"
                                              timeout={200}
                                              usePortal={true}
                                              zIndex={1500}
                                            />
                                          </Tooltip>
                                        </Tip>
                                      </div>
                                    </Box>
                                  </Flex>
                                </div>
                              </Box>
                            </Flex>
                          </ForwardRef>
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    className="simplebar-placeholder"
                  />
                </div>
                <div
                  className="simplebar-track simplebar-horizontal"
                >
                  <div
                    className="simplebar-scrollbar"
                  />
                </div>
                <div
                  className="simplebar-track simplebar-vertical"
                >
                  <div
                    className="simplebar-scrollbar"
                  />
                </div>
              </div>
            </SimpleBar>
          </InnerDock>
        </div>
      </Box>
    </Flex>
  </ForwardRef>
</Memo(Variants)>
`;
